oxcache/src/device.rs

```diff
@@ -1,49 +1,263 @@
+use nvme::info::is_zoned_device;
+use nvme::ops::{zns_append, zns_read};
+use nvme::types::ZNSConfig;
+use std::collections::VecDeque;
 use std::fs::OpenOptions;
+use std::io::Error;
+use std::ops::Deref;
 use std::os::fd::{AsRawFd, RawFd};
-use std::sync::Arc;
+use std::sync::{Arc, Mutex};
+
 use crate::cache::bucket::ChunkLocation;
 use crate::device;
+use crate::eviction::EvictionPolicy;
+
+#[derive(Copy, Clone)]
+struct Zone {
+    index: usize,
+    chunks_available: usize,
+}
+
+struct ZoneList {
+    available_zones: VecDeque<Zone>,
+    chunks_per_zone: usize,
+}
+
+impl ZoneList {
+    fn new(num_zones: usize, chunks_per_zone: usize) -> Self {
+        let mut avail_zones = VecDeque::with_capacity(num_zones as usize);
+
+        for i in 0..num_zones {
+            avail_zones.push_back(Zone {
+                index: i,
+                chunks_available: chunks_per_zone,
+            });
+        }
+
+        ZoneList {
+            available_zones: avail_zones,
+            chunks_per_zone,
+        }
+    }
+
+    // Get a zone to write to
+    fn remove(&mut self) -> Result<usize, ()> {
+        if self.is_full() {
+            // Need to evict
+            return Err(());
+        }
+
+        let mut zone = self.available_zones.pop_front().unwrap();
```

Comment:
Shouldn't we use ?

```Rust
let mut zone = self.available_zones.pop_front()?;
```

Or handle the error here so we don't panic in case it's empty?